aerogear/graphback
# Generate unidirectional fields in generated schema

## Motivation

https://github.com/aerogear/graphback/issues/614

## What

This change uses a one-to-many field in the data model to generate the related field in the opposing type and inputs.

Add generated fields to:

- types
- filters
- inputs 

## Why

This enables the developer to perform mutations and queries from both types with minimal configuration.

## Verification Steps

Set the following data model:

```gql
type User {
  id: ID!
  name: String!
  """
  @db.oneToMany: 'inUser'
  """
  inTransactions: [Transaction!]!
  """
  @db.oneToMany: 'outUser'
  """
  outTransactions: [Transaction!]!
}

type Transaction {
  id: ID!
  amount: Int!
  outUser: User
}
```

Run `graphback generate` and `graphback db` to generate your code and migrate your database structure.

Look at your generated schema in `generated.ts`. The `Transaction` type should have automatically added the `inUser` field.

```gql
type Transaction {
  id: ID!
  amount: Int!
  inUser: User!
  outUser: User
}
```

Run the following mutations and queries in order to verify functionality.

```gql
mutation createUserA {
  createUser(input:{name:"User A"}) {
    id
    name
  }
}

mutation createUserB {
  createUser(input:{name:"User B"}) {
    id
    name
  }
}

mutation createTransaction {
  createTransaction(input: {amount:100, inUserId:1, outUserId:2}) {
    id
    amount
    inUser {
      id
      name
    }
    outUser {
      id
      name
    }
  }
}

query findTransactions {
  findAllTransactions {
    id
    amount
    inUser {
      id
      name
    }
    outUser {
      id
      name
    }
  }
}

query getUsersAndTransactions {
  findAllUsers {
    id
    name
    inTransactions {
      id
      amount
    }
    outTransactions {
      id
      amount
    }
  }
}
```
